Make error: '_locale_t' does not name a type

#### System Information
- Ogre Version: 1.10.10
- Operating System / Platform: Win7 64bit MinGW
- RenderSystem: OpenGL

#### Detailled description
Error occured when make: 

D:/Documents/Downloads/ogre-1.10.10/OgreMain/include/OgreString.h:61:21: error:
'_locale_t' does not name a type

--- a/CMake/Dependencies.cmake
+++ b/CMake/Dependencies.cmake
@@ -97,7 +97,7 @@ if(OGRE_BUILD_DEPENDENCIES AND NOT EXISTS ${OGREDEPS_PATH})
 
     set(BUILD_COMMAND_OPTS --target install --config ${CMAKE_BUILD_TYPE})
 
-    if(MSVC OR EMSCRIPTEN) # other platforms ship zlib
+    if(MSVC OR EMSCRIPTEN OR MINGW) # other platforms ship zlib
         message(STATUS "Building zlib")
         file(DOWNLOAD 
             http://zlib.net/zlib-1.2.11.tar.gz
@@ -163,7 +163,7 @@ if(OGRE_BUILD_DEPENDENCIES AND NOT EXISTS ${OGREDEPS_PATH})
     execute_process(COMMAND ${CMAKE_COMMAND}
         --build ${CMAKE_BINARY_DIR}/freetype-2.9/objs ${BUILD_COMMAND_OPTS})
 
-    if(MSVC) # other platforms dont need this
+    if(MSVC OR MINGW) # other platforms dont need this
         message(STATUS "Building SDL2")
         file(DOWNLOAD
             https://libsdl.org/release/SDL2-2.0.8.tar.gz

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -160,8 +160,6 @@ if (MINGW)
     # the atomic primitives
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=i686")
   endif ()
-  # Fpermissive required to let some "non-standard" casting operations in the plugins pass
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fpermissive")
   # disable this optimisation because it breaks release builds (reason unknown)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-tree-slp-vectorize")
   # Ignore some really annoying warnings which also happen in dependencies

--- a/OgreMain/include/OgreColourValue.h
+++ b/OgreMain/include/OgreColourValue.h
@@ -318,7 +318,7 @@ namespace Ogre {
 
         /** Function for writing to a stream.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const ColourValue& c )
         {
             o << "ColourValue(" << c.r << ", " << c.g << ", " << c.b << ", " << c.a << ")";

--- a/OgreMain/include/OgreMath.h
+++ b/OgreMain/include/OgreMath.h
@@ -82,7 +82,7 @@ namespace Ogre
         bool operator >= ( const Radian& r ) const { return mRad >= r.mRad; }
         bool operator >  ( const Radian& r ) const { return mRad >  r.mRad; }
 
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Radian& v )
         {
             o << "Radian(" << v.valueRadians() << ")";
@@ -133,7 +133,7 @@ namespace Ogre
         bool operator >= ( const Degree& d ) const { return mDeg >= d.mDeg; }
         bool operator >  ( const Degree& d ) const { return mDeg >  d.mDeg; }
 
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Degree& v )
         {
             o << "Degree(" << v.valueDegrees() << ")";

--- a/OgreMain/include/OgreMatrix3.h
+++ b/OgreMain/include/OgreMatrix3.h
@@ -227,7 +227,7 @@ namespace Ogre
 
         /** Function for writing to a stream.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Matrix3& mat )
         {
             o << "Matrix3(" << mat[0][0] << ", " << mat[0][1] << ", " << mat[0][2] << "; "

--- a/OgreMain/include/OgreMatrix4.h
+++ b/OgreMain/include/OgreMatrix4.h
@@ -172,7 +172,7 @@ namespace Ogre
 
         /** Function for writing to a stream.
          */
-        inline _OgreExport friend std::ostream& operator<<(std::ostream& o, const TransformBase& mat)
+        inline friend std::ostream& operator<<(std::ostream& o, const TransformBase& mat)
         {
             o << "Matrix4(";
             for (size_t i = 0; i < 4; ++i)

--- a/OgreMain/include/OgreQuaternion.h
+++ b/OgreMain/include/OgreQuaternion.h
@@ -329,7 +329,7 @@ namespace Ogre {
         /** Function for writing to a stream. Outputs "Quaternion(w, x, y, z)" with w,x,y,z
             being the member values of the quaternion.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Quaternion& q )
         {
             o << "Quaternion(" << q.w << ", " << q.x << ", " << q.y << ", " << q.z << ")";

--- a/OgreMain/include/OgreVector2.h
+++ b/OgreMain/include/OgreVector2.h
@@ -506,7 +506,7 @@ namespace Ogre
 
         /** Function for writing to a stream.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Vector2& v )
         {
             o << "Vector2(" << v.x << ", " << v.y <<  ")";

--- a/OgreMain/include/OgreVector3.h
+++ b/OgreMain/include/OgreVector3.h
@@ -744,7 +744,7 @@ namespace Ogre
 
         /** Function for writing to a stream.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Vector3& v )
         {
             o << "Vector3(" << v.x << ", " << v.y << ", " << v.z << ")";

--- a/OgreMain/include/OgreVector4.h
+++ b/OgreMain/include/OgreVector4.h
@@ -370,7 +370,7 @@ namespace Ogre
         }
         /** Function for writing to a stream.
         */
-        inline _OgreExport friend std::ostream& operator <<
+        inline friend std::ostream& operator <<
             ( std::ostream& o, const Vector4& v )
         {
             o << "Vector4(" << v.x << ", " << v.y << ", " << v.z << ", " << v.w << ")";

--- a/OgreMain/src/OgreStringConverter.cpp
+++ b/OgreMain/src/OgreStringConverter.cpp
@@ -27,12 +27,12 @@ THE SOFTWARE.
 */
 #include "OgreStableHeaders.h"
 
-#if OGRE_PLATFORM == OGRE_PLATFORM_WIN32 || OGRE_PLATFORM == OGRE_PLATFORM_WINRT
+#if (OGRE_PLATFORM == OGRE_PLATFORM_WIN32 || OGRE_PLATFORM == OGRE_PLATFORM_WINRT) && !defined(__MINGW32__)
 #   define LC_NUMERIC_MASK LC_NUMERIC
 #   define newlocale(cat, loc, base) _create_locale(cat, loc)
 #endif
 
-#if OGRE_PLATFORM == OGRE_PLATFORM_ANDROID || OGRE_PLATFORM == OGRE_PLATFORM_EMSCRIPTEN
+#if OGRE_PLATFORM == OGRE_PLATFORM_ANDROID || OGRE_PLATFORM == OGRE_PLATFORM_EMSCRIPTEN || defined(__MINGW32__)
 #   define newlocale(cat, loc, base) 0
 #endif
 

--- a/RenderSystems/Direct3D9/include/OgreMinGWSupport.h
+++ b/RenderSystems/Direct3D9/include/OgreMinGWSupport.h
@@ -43,8 +43,7 @@ THE SOFTWARE.
 #   define __in
 #endif
 #define UINT8 uint8_t
-#define WINAPI_INLINE inline
-#ifndef __uuidof(Object)
+#ifndef __uuidof
 #   define __uuidof(Object) IID_##Object
 #endif
 

--- a/RenderSystems/Direct3D9/src/OgreD3D9Plugin.cpp
+++ b/RenderSystems/Direct3D9/src/OgreD3D9Plugin.cpp
@@ -33,7 +33,6 @@ THE SOFTWARE.
 extern "C" {   
 #include "OgreMinGWSupport.h"
 void _chkstk();
-void _fastcall __security_check_cookie(intptr_t i);
 }   
 #endif
 

--- a/RenderSystems/Direct3D9/src/OgreDxErr.cpp
+++ b/RenderSystems/Direct3D9/src/OgreDxErr.cpp
@@ -65,11 +65,14 @@ A "contributor" is any person that distributes its contribution under this licen
 
 // This version only supports UNICODE.
 
+#include "OgrePrerequisites.h"
+
 #if !defined(NOMINMAX)
 #define NOMINMAX
 #endif
 #include <windows.h>
 
+#ifndef __MINGW32__
 #pragma warning( disable : 4005 ) // warning C4005: 'xxx': macro redefinition
 
 #include <stdio.h>
@@ -4025,11 +4028,12 @@ HRESULT WINAPI DXTraceW( _In_z_ const WCHAR* strFile, _In_ DWORD dwLine, _In_ HR
 
     return hr;
 }
-
+#endif
 namespace Ogre
 {
 	String DXGetErrorDescription(HRESULT hr)
 	{
+#ifndef __MINGW32__
 		WCHAR wideBuf[1024];
 		wideBuf[0] = 0;
 		DXGetErrorDescriptionW(hr, wideBuf, ARRAYSIZE(wideBuf));
@@ -4040,5 +4044,8 @@ namespace Ogre
 			narrowBuf[0] = 0;
 
 		return String(narrowBuf);
+#else
+		return "";
+#endif
 	}
 }

--- a/RenderSystems/Direct3D9/src/OgreMinGWSupport.cpp
+++ b/RenderSystems/Direct3D9/src/OgreMinGWSupport.cpp
@@ -44,9 +44,7 @@ extern "C" {
     
     // MSVC uses security cookies to prevent some buffer overflow attacks.
     // provide dummy implementations.
-    intptr_t __security_cookie;
-    
-    void _fastcall __security_check_cookie(intptr_t i)
+    void __security_check_cookie(intptr_t i)
     {
     }
 }

--- a/RenderSystems/GLSupport/src/win32/OgreWin32GLSupport.cpp
+++ b/RenderSystems/GLSupport/src/win32/OgreWin32GLSupport.cpp
@@ -441,11 +441,11 @@ namespace Ogre {
 
     void* Win32GLSupport::getProcAddress(const char* procname)
     {
-        void* res = wglGetProcAddress(procname);
+        void* res = (void*)wglGetProcAddress(procname);
 
         if (!res) {
             static HMODULE libgl = LoadLibraryA("opengl32.dll");
-            res = GetProcAddress(libgl, procname);
+            res = (void*)GetProcAddress(libgl, procname);
         }
 
         return res;

